### PR TITLE
Generate MapKey creators

### LINKED ISSUE #651: FR: Generate MapKeyCreator classes in factory gen
When using `@MapKey(unwrapValue = false)`, dagger generates creator classes for annotations that are consumed in components to create annotation instances. It would be useful to generate these in the signature Dagger expects as well.

```kotlin
// Given this
@MapKey(unwrapValue = false)
annotation class MultiScopeActivityKey(
  val value: KClass<out Activity>,
  val scope: KClass<*>,
)

// Generate this
object MultiScopeActivityKeyCreator {
  @JvmStatic
  fun createMultiScopeActivityKey(
    value: Class<out Activity>,
    scope: Class<*>
  ): MultiScopeActivityKey {
    return MultiScopeActivityKey(value.kotlin, scope.kotlin)
  }
}
```

The nice thing with Kotlin 1.7 is we can instantiate annotations directly, which is simpler than Dagger's standard mechanism of using AutoAnnotation

### PR DESCRIPTION
**Pull request by ZacSweers**
Resolves #651

### CODE REVIEW

File: compiler-utils/build.gradle [hunk 1]
```diff
@@ -21,6 +21,8 @@ dependencies {
   testFixturesImplementation project(':compiler')
   testFixturesImplementation deps.dagger2.compiler
   testFixturesImplementation deps.dagger2.dagger
+  testFixturesImplementation deps.auto.value.annotations
+  testFixturesImplementation deps.auto.value.processor
```
Comment (ZacSweers) [inline compiler-utils/build.gradle:25]: Necessary because this is what dagger uses when it runs to support instantiating annotations at runtime

**Comment by ZacSweers** [comment]
Hmm, this won't work with the old backend since instantiating annotations is only implemented in the new backend

**Comment by vRallev** [comment]
> Hmm, this won't work with the old backend since instantiating annotations is only implemented in the new backend

That sucks. [Seems like](https://youtrack.jetbrains.com/issue/KT-53135) this has been pushed out to Kotlin 1.8.20. See also #422. Until then it seems to be blocked.

**Comment by ZacSweers** [comment]
Is that actually a blocker since this only runs in factory generation?

**Comment by vRallev** [comment]
What if you enabled KAPT in the same module, because you have a `@Component`... In this case you should not enable factory generation anyways 🙃 You're right.

**Review by JoelWilcox — APPROVED** [review]
I would like to understand the annotation visiting chunk better but otherwise LGTM

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    val creatorsToGenerator = mutableSetOf<ClassReference>()
+
+    fun visitAnnotations(clazz: ClassReference) {
+      if (clazz.isAnnotationClass()) {
+        val added = creatorsToGenerator.add(clazz)
+        if (added) {
+          for (property in clazz.properties) {
+            val type = property.type().asClassReferenceOrNull()
+            if (type?.isAnnotationClass() == true) {
+              visitAnnotations(type)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt:125]: Annotation properties can be other annotations. It appears that dagger generates creators for every used annotation along with the original mapkey annotation itself.

So for this example

```kotlin
annotation class ValueHolder(val name: String)

@MapKey
annotation class CustomMapKey(val value: ValueHolder)

@CustomMapKey(ValueHolder("foo"))
class ActualInjectedType @Inject constructor(...)
```

Dagger then expects something like this

```java
public final class CustomMapKeyCreator {
  public static CustomMapKey createCustomMapKey(ValueHolder value) {
    // ...
  }

  // Also other used annotations
  public static ValueHolder createValueHolder(String value) {
    // ...
  }
}
```

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    if (!clazz.isAnnotationClass()) {
+      throw AnvilCompilationExceptionClassReference(
+        message = "@MapKey is only applicable to annotation classes.",
+        classReference = clazz
+      )
+    }
+
+    val className = clazz.asClassName()
+
+    val creatorsToGenerator = mutableSetOf<ClassReference>()
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt]: 321bb23

File: compiler-utils/build.gradle (hunk 1 above)
Comment (ZacSweers) [reply compiler-utils/build.gradle]: Good idea - 888cc04

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoelWilcox) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MapKeyCreatorGenerator.kt:125]: Thanks for explaining!

**Pull request merged**